### ISSUE TITLE
MYSQL_ROOT_PASSWORD_FILE ignored during docker stack deploy

### ISSUE THREAD

**Issue by SoftcatServiceDesk** [docker-library/mysql#501]
Attempting to use the following docker compose file for a stack results in an error:
````
version: '3.7'

services:
  db:
    image: mysql
    secrets:
      - db_password
    environment:
      - MYSQL_DATABASE=cm
      - MYSQL_ROOT_PASSWORD_FILE=/run/secrets/db_password

  adminer:
    image: adminer
    ports:
      - 8080:8080
    depends_on:
      - db

secrets:
  db_password:
    external: true
````

Error within the service logs (Full logs as comment):
```
2018-10-05T14:23:54.293125Z 5 [Warning] [MY-010453] [Server] root@localhost is created with an empty password ! Please consider switching off the --initialize-insecure option
```

Performing `docker exec -it container_name echo /run/secrets/db_password` echo's the password correctly?

**Comment by SoftcatServiceDesk** [comment]
```
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | Initializing database
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:47.581393Z 0 [Warning] [MY-011070] [Server] 'Disabling symbolic links using --skip-symbolic-links (or equivalent) is the default. Consider not using this option as it' is deprecated and will be removed in a future release.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:47.581503Z 0 [System] [MY-013169] [Server] /usr/sbin/mysqld (mysqld 8.0.12) initializing of server in progress as process 30
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | mbind: Operation not permitted
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:54.293125Z 5 [Warning] [MY-010453] [Server] root@localhost is created with an empty password ! Please consider switching off the --initialize-insecure option.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:57.523237Z 5 [Warning] [MY-010315] [Server] 'user' entry 'mysql.infoschema@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:57.523488Z 5 [Warning] [MY-010315] [Server] 'user' entry 'mysql.session@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:57.523616Z 5 [Warning] [MY-010315] [Server] 'user' entry 'mysql.sys@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:57.523791Z 5 [Warning] [MY-010315] [Server] 'user' entry 'root@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:57.524021Z 5 [Warning] [MY-010323] [Server] 'db' entry 'performance_schema mysql.session@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:57.524185Z 5 [Warning] [MY-010323] [Server] 'db' entry 'sys mysql.sys@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:57.524324Z 5 [Warning] [MY-010311] [Server] 'proxies_priv' entry '@ root@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:57.524800Z 5 [Warning] [MY-010330] [Server] 'tables_priv' entry 'user mysql.session@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:57.524939Z 5 [Warning] [MY-010330] [Server] 'tables_priv' entry 'sys_config mysql.sys@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:23:58.639657Z 0 [System] [MY-013170] [Server] /usr/sbin/mysqld (mysqld 8.0.12) initializing of server has completed
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | Database initialized
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | MySQL init process in progress...
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:00.579429Z 0 [Warning] [MY-011070] [Server] 'Disabling symbolic links using --skip-symbolic-links (or equivalent) is the default. Consider not using this option as it' is deprecated and will be removed in a future release.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:00.581066Z 0 [System] [MY-010116] [Server] /usr/sbin/mysqld (mysqld 8.0.12) starting as process 81
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | mbind: Operation not permitted
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | MySQL init process in progress...
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:01.934324Z 0 [Warning] [MY-010068] [Server] CA certificate ca.pem is self signed.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:01.945999Z 0 [Warning] [MY-011810] [Server] Insecure configuration for --pid-file: Location '/var/run/mysqld' in the path is accessible to all OS users.
Consider choosing a different directory.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:02.060453Z 0 [Warning] [MY-010315] [Server] 'user' entry 'mysql.infoschema@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:02.061290Z 0 [Warning] [MY-010315] [Server] 'user' entry 'mysql.session@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:02.061390Z 0 [Warning] [MY-010315] [Server] 'user' entry 'mysql.sys@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:02.061476Z 0 [Warning] [MY-010315] [Server] 'user' entry 'root@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:02.061577Z 0 [Warning] [MY-010323] [Server] 'db' entry 'performance_schema mysql.session@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:02.061662Z 0 [Warning] [MY-010323] [Server] 'db' entry 'sys mysql.sys@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:02.061745Z 0 [Warning] [MY-010311] [Server] 'proxies_priv' entry '@ root@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:02.082476Z 0 [Warning] [MY-010330] [Server] 'tables_priv' entry 'user mysql.session@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:02.082590Z 0 [Warning] [MY-010330] [Server] 'tables_priv' entry 'sys_config mysql.sys@localhost' ignored in --skip-name-resolve mode.
cm_db.1.nnzwmx89e96k@linuxkit-00155d94ed38    | 2018-10-05T14:24:02.158134Z 0 [System] [MY-010931] [Server] /usr/sbin/mysqld: ready for connections. Version: '8.0.12'  socket: '/var/run/mysqld/mysqld.sock'  port: 0  MySQL Community Server - GPL.
```

**Comment by wglambert** [comment]
https://docs.docker.com/engine/swarm/secrets/#use-secrets-in-compose
You put 
```yml
secrets:
  db_password:
    external: true
```
Which should be
```yml
secrets:
  db_password:
    file: db_password
```

The `MYSQL_ROOT_PASSWORD_FILE=/run/secrets/db_password` lists the directory to be mounted in the container for the secret, and the `file: db_password` is your local path to the file, relative to the build context.

```console
$ docker stack deploy -c docker-compose.yml mysql
Creating network mysql_default
Creating secret mysql_db_password
Creating service mysql_db
Creating service mysql_adminer

$ docker exec -it mysql_db.1.hfgqijvwlxkw1wqgd004jk3qc bash

root@ff91d5c36de0:/# env
HOSTNAME=ff91d5c36de0
MYSQL_DATABASE=cm
MYSQL_ROOT_PASSWORD_FILE=/run/secrets/db_password
PWD=/
HOME=/root
MYSQL_MAJOR=8.0
GOSU_VERSION=1.7
MYSQL_VERSION=8.0.12-1debian9
TERM=xterm
SHLVL=1
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
_=/usr/bin/env

root@ff91d5c36de0:/# mysql
Welcome to the MySQL monitor.  Commands end with ; or \g.
Your MySQL connection id is 10
Server version: 8.0.12 MySQL Community Server - GPL

Copyright (c) 2000, 2018, Oracle and/or its affiliates. All rights reserved.

Oracle is a registered trademark of Oracle Corporation and/or its
affiliates. Other names may be trademarks of their respective
owners.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

mysql>
```

**Comment by SoftcatServiceDesk** [comment]
Even with that configuration it still does the same thing - As I mentioned the secret is loaded successfully by the container (As per the following command: docker exec -it container_name echo /run/secrets/db_password). I believe the issue is perhaps between docker for windows and the mysql docker image

**Comment by SoftcatServiceDesk** [comment]
Further update - I think the issue lies with creating secrets like this on windows:
`echo password | docker secret create db_password`

For future reference - Create a file first and then use `docker secret create db_password db_password.txt` which works fine

**Comment by mariotacke** [comment]
Thank you @SoftcatServiceDesk, this was driving me crazy. I figured out it had to do with the line endings of the secret when created on Windows. I'm running cmder/cygwin as my shell and used the following to create a password on Windows:

```
openssl rand -base64 20 | docker secret create mysql-root_password -
```

After the database initialized, I was not able to log in with the new password. With your suggestion, I created a file first instead:

```
openssl rand -base64 20 >> password.txt
```

Then I inspected the newly generated file with `vim -b password.txt` (note the `-b` to see binary/markers):

It showed:
```
KTlxpVwttDsYiC9/9XFhTUHT9ls=^M
```

Removing the `^M` at the end and then creating a secret fixed that for me:
```
docker secret create mysql-root_password password.txt
```
